src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+            string explicitInterface = null;
+            foreach (var current in collectionType.AllInterfaces)
+            {
+                if (!knownCollectionInterfaces.Contains(current.OriginalDefinition))
+                {
+                    continue;
+                }
+
+                // see how the type implements the interface
+                var countSymbol = GetInterfaceMembers(current, get_Count);
+                var indexerSymbol = GetInterfaceMembers(current, get_Item);
+                if (countSymbol == null || indexerSymbol == null)
+                {
+                    continue;
+                }
+
+                var countImpl = collectionType.FindImplementationForInterfaceMember(countSymbol) as IMethodSymbol;
+                var indexerImpl = collectionType.FindImplementationForInterfaceMember(indexerSymbol) as IMethodSymbol;
+                if (countImpl == null || indexerImpl == null)
+                {
+                    continue;
+                }
+
+                if (!semanticFact.IsAssignableTo(indexerImpl.ReturnType, foreachType, model.Compilation))
+                {
+                    continue;
+                }
+
+                // implicitly implemented!
+                if (countImpl.ExplicitInterfaceImplementations.IsEmpty &&
+                    indexerImpl.ExplicitInterfaceImplementations.IsEmpty)
+                {
+                    explicitCast = null;
+                    countName = Count;
+                    return;
+                }
+
+                if (explicitInterface == null)
+                {
+                    explicitInterface = current.ToDisplayString();
+                }
+            }
+
+            // okay, we don't have implicitly implemented one, but we do have explicitly implemented one
+            if (explicitInterface != null)
+            {
+                explicitCast = explicitInterface;
+                countName = Count;
+            }
+        }
+
+        private static bool IsNullOrErrorType(ITypeSymbol type)
+        {
+            return type == null || type is IErrorTypeSymbol;
+        }
+
+        private static IMethodSymbol GetInterfaceMembers(ITypeSymbol interfaceType, string memberName)
+        {
+            var members = interfaceType.GetMembers(memberName);
+            if (!members.IsEmpty)
+            {
+                return (IMethodSymbol)members[0];
+            }
+
+            foreach (var current in interfaceType.AllInterfaces)
+            {
+                var member = GetInterfaceMembers(current, memberName);
+                if (member != null)
+                {
+                    return member;
+                }
+            }
+
+            return null;
+        }
+
+        private static bool CheckRequireCollectionStatement(IOperation operation)
+        {
+            return operation.Kind != OperationKind.LocalReference && operation.Kind != OperationKind.FieldReference;
```

Comment:
what about parameters? (and add test for this please).

Reply:
yes. parameter should be included.